### ISSUE TITLE
Consider the time period between card creation date and first review date in S0 calculation

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#713]
Consider three cards:
- Card 1: Created and rated `Good` today
- Card 2: Created 1 week ago and rated `Good` today
- Card 3: Created 1 month ago and rated `Good` today

Currently, FSRS assigns the same S0 to all of these cards. But, if we can find a way to use the time period between card creation date and first review date in calculating S0, we can assign a higher S0 to Card 3 than Card 2 and Card 1. 

Obviously, this feature will be useful only for people who make their own cards and not for those who use pre-made decks. But still...

**Comment by Expertium** [comment]
I don't think that's possible
1) As far as I know, in Anki we can't access the card creation date during optimization anyway
2) Even if we can, we can't distinguish between pre-made cards and cards made by the user himself
3) I can't think of any meaningful way to incorporate this into S0

**Comment by brishtibheja** [comment]
After spending this year in the forums, I think most users make their own cards rather than get shared decks.

Just throwing an idea out there, note modification date might be better than card creation? You can modify a note to change the flashcard's formulation. Or you may cloze some additional parts of the note to create new cards.

**Comment by user1823** [comment]
> As far as I know, in Anki we can't access the card creation date during optimization anyway

card creation date is deduced from cid, which is obviously available

> we can't distinguish between pre-made cards and cards made by the user himself

In general, the time difference would be much larger for pre-made cards. Just for research purposes, one can filter out cards whose first review was done at least 1 year after card creation date. Once we get some idea on how to use the info, we can develop a better filter.

> I can't think of any meaningful way to incorporate this into S0

One crude idea:
Let (first revlog id - cid)/(86400 * 1000) = x

For each first rating, group cards by x.
In each group, calculate S0 just like we do it now.

For each first rating, find a curve that fits the S0 values for each value of x.

**Comment by user1823** [comment]
> Wouldn't that result in having to calculate tons of S0's for every possible x?

Yes, but after that, we can try to find a curve that fits most of them.

**Comment by Expertium** [comment]
> Yes, but after that, we can try to find a curve that fits most of them.

Can you write some Python code or pseudocode to make it easier to understand exactly what you have in mind?

**Comment by Expertium** [comment]
Alright, but I still don't think that this is better than just calculating S0 the way it's done right now.

**Comment by user1823** [comment]
Just to record my thoughts:

> I still don't think that this is better than just calculating S0 the way it's done right now.

When the user makes a card themselves and reviews it in Anki, they are not using Anki to _learn_ the information. Rather, they are using Anki to _retain_ what they have learnt from other sources.

So, if the first rating is Good, it means that they could successfully recall what they learnt earlier outside Anki. So, realistically speaking, we are calculating the next stability and not the initial stability.

Clearly, the next stability depends upon the previous stability (the one before the first rating). Unfortunately, we don't have a good way to measure the previous one. But, at least, we can say that the cards rated Good after 30 days will, on average, have a greater stability than the cards rated Good after 7 days.

Currently, FSRS treats these cards in the same way. Calculating the S0 separately after grouping the cards by number of days between card creation and first review will likely improve the calculation of S0.

<hr>

A quick and dirty way to test this idea:

- Group the cards by number of days between card creation and first review (x)
- Use the current formula to calculate S0 for each group and prepare a matrix containing S0 for each rating and x.
- Use this matrix to get the S0 for calculating the subsequent memory states.

Advantage over the previously proposed method:
- It can be done quickly (for testing); we don't need to spend time thinking about the relation between x and S0.

Disadvantages over the previously proposed method:
- Can't be used in Anki (the matrix is too large to be stored in Anki)
- Some groups will have low number of cards, which can make this more susceptible to noise. Using a mathematical relation (as previously proposed) will solve this issue.

**Comment by DerIshmaelite** [comment]
> When the user makes a card themselves and reviews it in Anki, they are not using Anki to learn the information. Rather, they are using Anki to retain what they have learnt from other sources.

That is not true. There are people (I for one) who are using Anki as the starting point from which people begin to acquire information.

**Comment by user1823** [comment]
If you are creating your own cards, you must first see that information outside Anki, right? It's ok if you don't know it that well but you still know it at the time you are creating the card.

Even if I my assumption about _your_ use-case is wrong, my point is that there are at least some (or maybe even many) people who Anki to retain the information learnt outside Anki.

I believe that my suggestion would significantly improve the accuracy of FSRS for them. Once we have it working, we can discuss how to identify the users who Anki in this way.

**Comment by Expertium** [comment]
Honestly, I highly doubt it will be worth it, and it will probably be very difficult to figure out how to calculate the new S0. So unless Jarrett wants to work on it, I don't think this will be done, since I don't see the point.

**Comment by user1823** [comment]
Since there was no response from @L-M-Sherlock, I decided to do some testing on my own collection. The results are interesting. @Expertium 

I used the following code in the debug console to find cards whose first review was within less than 10 days of creating the card. (copy the debug console output and paste in the browser window to search)

```py
from aqt.utils import showText

sql = """
SELECT
     cid
FROM 
    revlog AS r1
WHERE 
    ROUND((id - cid) / (86400 * 1000)) < 10
AND
    id = (SELECT MIN(id) FROM revlog AS r2 WHERE r1.cid = r2.cid)
"""

cids = (str(cid) for cid in mw.col.db.list(sql))
search_query = f'"cid:{",".join(cids)}"'
showText(search_query, copyBtn=True)
```

Then, I moved these cards into a new deck and gave it a separate preset. Then, optimized both decks (presets).

<hr>

<h2>Observations</h2>

Parameters of complete deck (before any of the above operations): 
1.4843, 5.2091, 34.2467, 85.0440, 7.2761, 0.9915, 3.9806, 0.0010, 1.9269, 0.3428, 1.4228, 1.7287, 0.0304, 0.4317, 2.3433, 0.0171, 6.0000, 0.9510, 0.6386

Evaluation: Log loss: 0.2098, RMSE(bins): 1.32%.

<h3>Deck 1 (x < 10d):</h3>

No. of reviews: ⁨67,467⁩ reviews
Parameters:
0.9310, 3.0256, 17.7261, 35.8606, 7.3785, 0.9810, 3.9037, 0.0010, 1.7643, 0.2812, 1.2616, 1.8370, 0.1155, 0.4837, 2.0510, 0.0000, 6.0000, 0.9777, 0.6662

Evaluation with new parameters: Log loss: 0.2221, RMSE(bins): 1.69%.
Evaluation with parameters of complete deck: Log loss: 0.2227, RMSE(bins): 1.75%.

<h3>Deck 2 (x ≥ 10d):</h3>

No. of reviews: ⁨79,443⁩ reviews
Parameters:
2.4860, 8.1461, 48.3187, 73.8801, 7.1577, 0.9967, 3.9325, 0.0010, 1.9119, 0.3130, 1.4081, 2.0449, 0.0167, 0.3757, 2.3958, 0.0000, 6.0000, 0.7905, 0.6943

Evaluation with new parameters: Log loss: 0.1981, RMSE(bins): 1.34%.
Evaluation with parameters of complete deck: Log loss: 0.1986, RMSE(bins): 1.47%.

<h2>Conclusion</h2>

- There is a significant difference in the initial stabilities between the two subdecks.
- Splitting up the deck improved the RMSE's of both the subdecks (by 3.5% and 9% respectively).

**Comment by fiah3** [comment]
Just commenting to say that I agree with @user1823 suggestions! Have you made any progress with this? Is there any way it could become an add-on? I make anki cards while I revise the material, and sometimes study the cards on the same day so 8 day intervals for a card I just created obviously makes no sense!

**Comment by jmadelaine** [comment]
Just commenting to say I disagree with @user1823. I, for one, create my own cards programmatically using CSV files and JS. Some of the decks I create have over 5000 cards. I don't think we should assume that card creation date has any relation to retention at all, because it doesn't.

**Comment by brishtibheja** [comment]
Most of the userbase don't manually create their cards from text/csv imports. Besides, such something can also be parameterised.

**Comment by user1823** [comment]
I am not saying that there is a relationship between creation date and retention for every user. But, at least for some users, there is a relationship. So, it would be great if we could use that information to provide better scheduling for such users. The other users (for whom there isn't such relation) will remain unaffected.

**Comment by jmadelaine** [comment]
We could possibly allow an opt-in setting that considers the creation date in the scheduling, but to be honest, if the user wanted to begin the review process for a specific card, they could just review it on the same day they created it. As long as creation date is not considered by default, I don't mind what gets decided.
